### PR TITLE
Improve dest.Calculate() and dest.GetMultiBestPath() performance

### PR DESCRIPTION
**Pull request by nplanelcisco**
Improve performance by 10% ish by caching path attributes via a path attributes hash if needed

```
Before
BenchmarkMultiPath/Benchmark_Calculate-12                 598702              1837 ns/op             376 B/op         14 allocs/op
BenchmarkMultiPath/Benchmark_GetMultiBestPath-12         3420961               343.5 ns/op             0 B/op          0 allocs/op

After
BenchmarkMultiPath/Benchmark_Calculate-12         	  637652	      1764 ns/op	     376 B/op	      14 allocs/op
BenchmarkMultiPath/Benchmark_GetMultiBestPath-12  	 4771671	       298.0 ns/op	       0 B/op	       0 allocs/op

### CODE REVIEW

File: internal/pkg/table/destination.go [hunk 1]
```diff
@@ -692,11 +692,11 @@ func compareByMED(path1, path2 *Path) *Path {
 		firstAS := func(path *Path) uint32 {
 			if asPath := path.GetAsPath(); asPath != nil {
 				for _, v := range asPath.Value {
-					segType := v.GetType()
 					asList := v.GetAS()
 					if len(asList) == 0 {
 						continue
 					}
+					segType := v.GetType()
 					switch segType {
```
Comment (fujita) [inline internal/pkg/table/destination.go]: `switch v.GetType() {`
?

File: internal/pkg/table/path.go [hunk 2]
```diff
@@ -1034,19 +1031,41 @@ func (lhs *Path) Equal(rhs *Path) bool {
...
 
+	lhsPathAttrs := lhs.GetPathAttrs()
+	rhsPathAttrs := rhs.GetPathAttrs()
+	if len(lhsPathAttrs) == 0 && len(rhsPathAttrs) == 0 {
+		return lhs.EqualBySourceAndPathID(rhs)
+	}
+
+	if len(lhsPathAttrs) != len(rhsPathAttrs) {
+		return false
+	}
```
Comment (fujita) [inline internal/pkg/table/path.go:1046]: Could you add a comment explaining why it's better to do these comparisons first?

Comment (nplanelcisco) [reply internal/pkg/table/path.go:1046]: done

**Comment by fujita** [comment]
The zebra-nht failure doesn't look like a fluke?

**Comment by nplanelcisco** [comment]
> The zebra-nht failure doesn't look like a fluke?

Seems a regression, I will take a look tomorrow

**Comment by nplanelcisco** [comment]
> > The zebra-nht failure doesn't look like a fluke?
> 
> Seems a regression, I will take a look tomorrow

@fujita it's fix now, you can review/approve

**Comment by fujita** [comment]
Thank you.
Just out of curiosity, what caused the zebra tests to fail?

**Comment by nplanelcisco** [comment]
> Thank you. Just out of curiosity, what caused the zebra tests to fail?

I forgot to invalidate the `path.hashAttr` if a `add/delAttr` is called. What cause the issue is that Metrics (MED) was wrong for the path, as value changed but not the hash, so compare would return the wrong path.

I will add some tests for corner case like this and other behavior check (GetListAttrs() in order, ...)

**Comment by fujita** [comment]
> > Thank you. Just out of curiosity, what caused the zebra tests to fail?
> 
> I forgot to invalidate the `path.hashAttr` if a `add/delAttr` is called. What cause the issue is that Metrics (MED) was wrong for the path, as value changed but not the hash, so compare would return the wrong path.

Understood.
 
> I will add some tests for corner case like this and other behavior check (GetListAttrs() in order, ...)

Great, there should be tests other than the zebra tests that can catch that bug. Thanks!
